Cosmic drift update

sophiemiddleton opened the pull request.

brownd1978 commented on CosmicReco/src/CosmicMuonInfo_module.cc:
```diff
@@ -0,0 +1,431 @@
+
+mu2e::CosmicMuonInfo::CosmicMuonInfo(fhicl::ParameterSet const& pset):
+  art::EDFilter{pset},
+  _strawHitsTag(pset.get<std::string>("strawHitsTag")),
```
Please update this module to use validated fhicl

sophiemiddleton commented on CosmicReco/src/CosmicFitDisplay_module.cc:
```diff
@@ -0,0 +1,1498 @@
+    CosmicFitDisplay::CosmicFitDisplay(fhicl::ParameterSet const& pset) :
+	art::EDAnalyzer(pset),
+	
+	_mcdiag		(pset.get<bool>("MCdiag",true)),
```
ok, will look at this now. I will update in my local repo and then push. This should hopefully avoid any issues like last time - fingers crossed

sophiemiddleton replied:
ah, ok, do you have an example of this validated fhicl way of implementing this. I cant seem to see any difference in the modules I'm looking at?

brownd1978 commented on Mu2eUtilities/src/ParametricFit.cc:
```diff
@@ -0,0 +1,277 @@
+
+       }
+
+	XYZVec PointToLineCA(XYZVec& point, XYZVec& starting_point, XYZVec& end_point){
```
This function should be its own class, PointLinePCA_XYZ to follow the existing naming convention in Mu2eUtilities.

brownd1978 commented on Mu2eUtilities/src/SConscript:
```diff
@@ -46,7 +49,9 @@ mainlib = helper.make_mainlib ( [ 'mu2e_ConditionsService_ConditionsService_serv
-                                  'boost_system'
+                                  'boost_system',
+                                  extrarootlibs,
+                                  babarlibs 
```
This dependency comes from files that don't really belong in Mu2eUtilities

brownd1978 commented on RecoDataProducts/inc/CosmicTrack.hh:
```diff
@@ -0,0 +1,333 @@
+   };
+namespace mu2e {
+  
+  class CosmicTrack{
```
Is there anything about this class that is specific to 'Cosmic'?  ie is it really just a straight track?
Diagnostic information should be kept separate from the actual data describing the fit.  Much of the content of the *Diag structs seems to be computable from the fit, do you need persistence for that or could it be function content.

sophiemiddleton replied:
I liked the accessbility of these directly, the way this is done currently underpins the routine so it would take many weeks to re-do.....are there strong opinions on this?

brownd1978 replied:
You can't put MC truth diagnostic info into a RecoDataProducts class.  The other embedded diagnostics can be kept if you really need them.  Eventually you will have to refactor this code.

sophiemiddleton replied:
Ok, maybe I should move cosmic track to "data product" then and not "recodataproduct" the truth info is currently used for the analysis purposes not in the fitting routines, I need to think this through some more as to were is best to put the truth info. The truth info is only filled if mcdiag is set in the fitter.....this would not be the case if we were looking at data for example and then these truth members would just be empty....I thought that would be ok but I guess not...I'll rethink

sophiemiddleton replied:
For now I have just removed the truth info. As this progresses, this class may change shape but for current purposes I think itis now ok.

brownd1978 commented on RecoDataProducts/inc/CosmicTrackSeed.hh:
```diff
@@ -0,0 +1,50 @@
+    
+    TrkT0	             _t0;	      // t0 for this track
+    ComboHitCollection       _panel_hits;	      // hits for track (panel hits)
+    ComboHitCollection       _straw_chits;    // get the straw level hits and store here (need to find panel hits first)
```
Is there a reason you are deep copying the ComboHits, and not just keeping indices as in TimeCluster?

sophiemiddleton replied:
investigating....I think there is a reason....

brownd1978 commented on RecoDataProducts/inc/CosmicTrkFitFlag.hh:
```diff
@@ -0,0 +1,26 @@
+#include <map>
+namespace mu2e {
+
+  struct CosmicTrkFitFlagDetail {
```
As we discussed the cosmic bit should be merged into TrkFitFlag.

sophiemiddleton replied:
added 1 bit to TrkFitFlag for "isStraight" then can reuse other flags. This was discussed at the CM

brownd1978 commented on RecoDataProducts/src/SConscript:
```diff
@@ -43,6 +45,7 @@ helper.make_dict_and_map( [ mainlib,
                             'mu2e_GeomPrimitives',
                             'mu2e_Mu2eInterfaces',
                             'mu2e_GeneralUtilities',
+                            'mu2e_TrackerGeom',
```
What brings in this dependency?

sophiemiddleton replied:
removed

brownd1978 commented on TrkDiag/inc/ComboHitInfo.hh:
```diff
@@ -4,6 +4,17 @@
   // info about each hit in the combo hit
   struct ComboHitInfo {
+    // my addition vvvvvvvvvv
+    Float_t _posx, _posy, _posz; // position of this hit
```
Please just embed an XYZVec instead of separate floats for positions, etc.

sophiemiddleton replied:
I will remove these additions now

brownd1978 commented on TrkDiag/inc/TimeClusterInfo.hh:
```diff
@@ -22,6 +22,9 @@ namespace mu2e {
     Float_t _minhtime, _maxhtime; // min and max cluster hit time
     Float_t _maxover; // max overlap with another time cluster this event
     XYZVec _pos; // average position of cluster
+    Float_t _posx; // average position, S Middleton	
```
Why do you repeat the contents of the XYZVec as separate floats?

sophiemiddleton replied:
this maybe unnecessary I think I will just remove these additions if the code doesnt throw any exceptions to it as I think they are not really needed

brownd1978 commented on TrkDiag/inc/TimeClusterInfo.hh:
```diff
@@ -64,6 +67,9 @@ namespace mu2e {
     Int_t	_nceclust; // # of conversion electron hits found in clusters
     Float_t	_time; // average time of CE hits (doesn't include drift!)
     XYZVec	_pos; // average position of cluster
+    Float_t _mcposx; // average position, S Middleton
```
Please use XYZVec instead of separate floats

sophiemiddleton replied:
will either do or remove if not needed any more once I rejig above

sophiemiddleton commented:
Hi Can you point me in the direction of an example for the validated fhicl way of doing things, all the modules I see seem to use the same systems as I am using

brownd1978 commented:
Hi Sophie,
Andy included this in one of his tutorials.  There are lots of examples in
the code base, just look for struct Config.
https://mu2ewiki.fnal.gov/wiki/Module_Writing_Tutorial#Exercise_2:_Adding_module_configuration_.28Hello.2C_Fhicl_Validation.21.29

Dave

On Mon, Oct 28, 2019 at 10:24 AM Sophie Middleton <notifications@github.com>
wrote:

> Hi Can you point me in the direction of an example for the validated fhicl
> way of doing things, all the modules I see seem to use the same systems as
> I am using
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/40?email_source=notifications&email_token=ABAH575ZZTH4A3OMJESN2C3QQ4N5JA5CNFSM4JFGSWLKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECNWX5Q#issuecomment-547056630>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH5735T7AHJNA72TFU7CLQQ4N5JANCNFSM4JFGSWLA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

sophiemiddleton commented on CosmicReco/inc/CosmicTrackFinderData.hh:
```diff
@@ -0,0 +1,144 @@
+    
+    //possible use for alignment so keep for now:
+    struct ChannelID {
+      int Station;
```
will remove

sophiemiddleton replied:
ahh, it turns out this is used for the ordering of the hits so is needed for the fitting routine....is it ok to leave for now. This is the same way its done in helix finder, I didnt think it would be a bad idea to mirror

sophiemiddleton commented on CosmicReco/src/CosmicTrackFit.cc:
```diff
@@ -0,0 +1,654 @@
+  2) Applky my fit routine to the MC StrawDigiMC i.e. a way to assess accuracy of fitting routine when compared to reco.
+
+//----------------------------------------------------*/
+void CosmicTrackFit::FitMC(CosmicTrackFinderData& trackData, CosmicTrack* cosmictrack, bool Det){	
```
they dont, this is a separate function, you mean I should have separate CosmicTrackFitMC.cc

brownd1978 replied:
Yes.  Reconstruction classes cannot reference MC truth.  It may be you can run your fit unchanged on data provided by MC truth, then all you need is code to stuff the MC truth values into the right data structures.

sophiemiddleton replied:
moved

sophiemiddleton commented on CosmicReco/inc/CosmicTrackFinder_types.hh:
```diff
@@ -0,0 +1,76 @@
+    struct Data_t {
+      const art::Run*               run;
+      const art::Event*               event;
+      fhicl::ParameterSet*            timeOffsets; //what are these?
```
this seems to underpin the entire algorithim. It was based on helix finder- I will remove the PArameter Set as that is used but is there a real need to remove the run and event  here. As I said this is done like this is robust helix finder and seems to work ok...?

sophiemiddleton replied:
removed this file and all references to it

sophiemiddleton commented on Mu2eUtilities/inc/LiklihoodFunctions.hh:
```diff
@@ -0,0 +1,63 @@
+		std::vector<double> bestfiterrors;
+		std::vector<double> bestfitcov;
+
+		std::vector<double> StartDOCAs;
```
Yeah the CosmicTrack.hh has number of structs. I thought this was ok for now but I will upgrade once we start alignment. I think associating the diagnostics directly with the track was fine as they will be used for calibration and alignment purposes and therefore having things like the DriftDiag and Diag lists easily accessbile would make things easier?

sophiemiddleton commented on Mu2eUtilities/inc/PDFFit.hh:
```diff
@@ -0,0 +1,82 @@
+    
+    //std::vector<double> time_residuals;
+  
+    ComboHitCollection chits;
```
will move to CosmicReco

sophiemiddleton commented on Mu2eUtilities/inc/DriftFitUtils.hh:
```diff
@@ -0,0 +1,49 @@
+
+using namespace mu2e;
+
+namespace DriftFitUtils{
```
will move

sophiemiddleton replied:
I moved this but not even though it builds after the build I get this error:


  Internal Error: no cycle found for node lib/libmu2e_Mu2eG4.so (<SCons.Node.FS.File object at 0x51e1c00>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_TrackerMC.so (<SCons.Node.FS.File object at 0x52d7430>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_EventGenerator.so (<SCons.Node.FS.File object at 0x220dca0>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_CalPatRec.so (<SCons.Node.FS.File object at 0x2199cb0>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_Validation.so (<SCons.Node.FS.File object at 0x5374d00>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_TrkDiag.so (<SCons.Node.FS.File object at 0x5333190>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_ParticleID.so (<SCons.Node.FS.File object at 0x52189a0>) in state pending
  lib/libmu2e_ConditionsService_ConditionsService_service.so -> lib/libmu2e_Mu2eUtilities.so -> lib/libmu2e_TrackerConditions.so -> lib/libmu2e_ConditionsService_ConditionsService_service.so
  Internal Error: no cycle found for node lib/libmu2e_TrkHitReco.so (<SCons.Node.FS.File object at 0x534fdd0>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_ExtinctionMonitorFNAL_Digitization.so (<SCons.Node.FS.File object at 0x22501e0>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_Print.so (<SCons.Node.FS.File object at 0x521c1f0>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_CosmicReco.so (<SCons.Node.FS.File object at 0x538a680>) in state pending
  lib/libmu2e_Mu2eUtilities.so -> lib/libmu2e_TrackerConditions.so -> lib/libmu2e_ConditionsService_ConditionsService_service.so -> lib/libmu2e_Mu2eUtilities.so
  Internal Error: no cycle found for node lib/libmu2e_BTrkData.so (<SCons.Node.FS.File object at 0x2194ed0>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_TrkExt.so (<SCons.Node.FS.File object at 0x534d6e0>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_EventDisplay.so (<SCons.Node.FS.File object at 0x2206700>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_CRVAnalysis.so (<SCons.Node.FS.File object at 0x2150e80>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_TrkReco.so (<SCons.Node.FS.File object at 0x535d380>) in state pending
  Internal Error: no cycle found for node lib/libmu2e_CutAndCountAnalysis.so (<SCons.Node.FS.File object at 0x21e59f0>) in state pending
  lib/libmu2e_TrackerConditions.so -> lib/libmu2e_ConditionsService_ConditionsService_service.so -> lib/libmu2e_Mu2eUtilities.so -> lib/libmu2e_TrackerConditions.so

File "/cvmfs/mu2e.opensciencegrid.org/artexternals/scons/v3_0_5/lib/python2.7/site-packages/SCons/Taskmaster.py", line 1074, in cleanup

There is no trackergeom now in the mu2eutilities Sconscript but it seems to complain about cyclic dependencies, I think?

I have done a clean build and still see this- do you know what could have caused this?

sophiemiddleton commented:
hi thanks, ok, I will get all the minor corrections done today, the rejigging of the Fcl parameters may take me some time as I will have to make sure I read up on it first. I have added comments above, let me know if you have any responses. I'll push those minor corrctions before getting on with the fcl changes

sophiemiddleton commented on Mu2eUtilities/inc/TwoLinePCA_XYZ.hh:
```diff
@@ -46,7 +46,7 @@ namespace mu2e {
     double dca()   const { return _dca; }
     double dca2d() const { return _dca2d; }
-
+    double ambig()   const { return _ambig; }
```
I think it might be as the outcome of cross product had to be double....I will check

sophiemiddleton replied:
will rename LRambig....as for the type I think it is due to the reason I give above as its equated to a cross product then / itself and get the sign.....I think....I will look into c++ to see if there's a more sophisticated way to do that

sophiemiddleton replied:
finally remembered to fix this

sophiemiddleton commented on CosmicReco/src/CosmicTrackFit.cc:
```diff
@@ -0,0 +1,654 @@
+	cosmictrack->set_initchisq_dofX(S.GetChi2X()/abs(DOF));
+	cosmictrack->set_initchisq_dof(S.GetTotalChi2()/abs(DOF));
+	for (size_t f2=0; f2<nHits; ++f2){
+         	if(isnan(cosmictrack->GetTrackDirection().Mag2()) == true) continue;     
```
a throw back to debugging days

sophiemiddleton commented on Mu2eUtilities/inc/ConvertXYZ.hh:
```diff
@@ -0,0 +1,42 @@
+	return vec;
+
+}
+XYZVec ConvertToXYZ(Hep3Vector vec1){
```
will remove

sophiemiddleton commented on Mu2eUtilities/inc/ConvertXYZ.hh:
```diff
@@ -0,0 +1,42 @@
+	return point;
+
+}
+Hep3Vector ConvertToHep3Vector(XYZVec vec1){
```
will remove

kutschke commented:
Hi Sophie,

   I recognize this comment is moot as you will be removing these functions in favor
of existing ones.  However it's important for future work.

  Never use the following at file scope in header files:

using namespace mu2e;               // example of a using declaration
using CLHEP::Hep3Vector;          // example of a using directive

It is OK in a .cc file but only *after* all includes.

The reason for not using them at file scope in header files is that your header 
files get included in other peoples code and your "using" can change the meaning
of their code. 

Similarly if you put  a using in your .cc file before a #include you are potentially
changing the meaning of that header file.  


  Rob


> On Oct 28, 2019, at 2:09 PM, Sophie Middleton <notifications@github.com> wrote:
> 
> @sophiemiddleton commented on this pull request.
> 
> In Mu2eUtilities/inc/ConvertXYZ.hh:
> 
> > +#include "GeometryService/inc/GeomHandle.hh"
> +#include "GeometryService/inc/DetectorSystem.hh"
> +
> +using namespace mu2e;
> +
> +HepPoint ConvertToHepPoint(XYZVec vec1){
> +	HepPoint point(vec1.x(),vec1.y(),vec1.z());
> +	return point;
> +
> +}
> +Hep3Vector ConvertToHep3Vector(XYZVec vec1){
> +	Hep3Vector vec(vec1.x(),vec1.y(),vec1.z());
> +	return vec;
> +
> +}
> +XYZVec ConvertToXYZ(Hep3Vector vec1){
> 
> will remove
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

sophiemiddleton commented on Mu2eUtilities/inc/ConvertXYZ.hh:
```diff
@@ -0,0 +1,42 @@
+
+}
+
+XYZVec ConvertToDetFrame(XYZVec vec){
```
will move this function into the fit function such that it doesnt need to call the external file at all

brownd1978 commented:
Hi Sophie,
DataProduct is more general than RecoDataProduct, not less, so CosmicTrack
cannot move there.  You need to separate reco and MC content when making
persistent classes.

mcdiag should also not be in the fitter, as it is a reco algorithm.


On Mon, Oct 28, 2019 at 12:45 PM Sophie Middleton <notifications@github.com>
wrote:

> *@sophiemiddleton* commented on this pull request.
> ------------------------------
>
> In RecoDataProducts/inc/CosmicTrack.hh
> <https://github.com/Mu2e/Offline/pull/40#discussion_r339757337>:
>
> > +   	std::vector<double> RecoAmbigs;
> +	std::vector<double> TrueAmbigs;
> +
> +   	std::vector<double> FinalResidualsX;
> +   	std::vector<double> FinalResidualsY;
> +	
> +	std::vector<double> FinalErrX;
> +	std::vector<double> FinalErrY;
> +	std::vector<double> FinalErrTot;
> +	
> +   	TrackDriftDiag();
> +
> +   };
> +namespace mu2e {
> +
> +  class CosmicTrack{
>
> Ok, maybe I should move cosmic track to "data product" then and not
> "recodataproduct" the truth info is currently used for the analysis
> purposes not in the fitting routines, I need to think this through some
> more as to were is best to put the truth info. The truth info is only
> filled if mcdiag is set in the fitter.....this would not be the case if we
> were looking at data for example and then these truth members would just be
> empty....I thought that would be ok but I guess not...I'll rethink
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/40?email_source=notifications&email_token=ABAH573O4YRYK3I6EIOE6QDQQ46LXA5CNFSM4JFGSWLKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJOUN5Y#discussion_r339757337>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH575INRQOKUALT6CFYOLQQ46LXANCNFSM4JFGSWLA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

sophiemiddleton commented on Mu2eUtilities/src/BuildMatrixSums.cc:
```diff
@@ -0,0 +1,246 @@
+#include "DataProducts/inc/XYZVec.hh"
+#include "Mu2eUtilities/inc/BuildMatrixSums.hh"
+
+BuildMatrixSums::BuildMatrixSums() {
```
done

sophiemiddleton replied:
have renamed.....might move to CosmicReco too

brownd1978 commented:
Hi Sophie,
whatever crosproduct() returns you can cast it to an int.  The data member
and accessor should not be double.
Dave

On Mon, Oct 28, 2019 at 2:04 PM Sophie Middleton <notifications@github.com>
wrote:

> *@sophiemiddleton* commented on this pull request.
> ------------------------------
>
> In Mu2eUtilities/inc/TwoLinePCA_XYZ.hh
> <https://github.com/Mu2e/Offline/pull/40#discussion_r339792060>:
>
> > @@ -46,7 +46,7 @@ namespace mu2e {
>      // Accessors for the 3d and 2d DCA.
>      double dca()   const { return _dca; }
>      double dca2d() const { return _dca2d; }
> -
> +    double ambig()   const { return _ambig; }
>
> will rename LRambig....as for the type I think it is due to the reason I
> give above as its equated to a cross product then / itself and get the
> sign.....I think....I will look into c++ to see if there's a more
> sophisticated way to do that
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/40?email_source=notifications&email_token=ABAH57Z4GPHKLDEMF44NLYLQQ5HWXA5CNFSM4JFGSWLKYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCJO7PJQ#discussion_r339792060>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH575YGH6VNX4IBHNJQUTQQ5HWXANCNFSM4JFGSWLA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

sophiemiddleton commented:
Hi,

Not sure how to reply to specific comments. I'm thinking through the MC thing. Basically the purpose of the MC fit is to look at differences i.e. those plots with dA, dB etc. These wont be used for data obviously the are for validating the fit and for presentations only. What I think I will do is remove all the MC stuff from CosmicTrackFit.cc . I will make a file called CosmicTrackMC.cc this will have those few functions which will be a copy and paste. This function can then be called in the fitter_module if mcdiag is set on in there. The. mcdiag will be removed from CosmicTrackFit.cc. 

As for CosmicTrack.hh I'm not sure how to associate the MC information with that. I guess it needs its own class despite being small. This would have to be associated to the reco product via run and event number in a way analogous to how we know what the strawdigimcs are relative to the combohit (obviously not a direct analogy but just a way to explain what I mean). Thus, I should create some new product CosmicMCTrack.hh......the question is then which directory would that belong in?

kutschke commented:
Thanks Sophie,

  When it's time to reintroduce using Truth info, please ask Dave and/or me
for design guidance.  We understand that you need it but we want to keep 
algorithms and data products that will run on data 100% MC free.   This adds
a few extra hops but you will still be able to do everything that you need to do.

 Rob

> On Oct 29, 2019, at 11:29 AM, Sophie Middleton <notifications@github.com> wrote:
> 
> @sophiemiddleton commented on this pull request.
> 
> In RecoDataProducts/inc/CosmicTrack.hh:
> 
> > +   	std::vector<double> RecoAmbigs;
> +	std::vector<double> TrueAmbigs;
> +
> +   	std::vector<double> FinalResidualsX;
> +   	std::vector<double> FinalResidualsY;
> +	 
> +	std::vector<double> FinalErrX;
> +	std::vector<double> FinalErrY; 
> +	std::vector<double> FinalErrTot;
> +	
> +   	TrackDriftDiag();
> +   
> +   };
> +namespace mu2e {
> +  
> +  class CosmicTrack{
> 
> For now I have just removed the truth info. As this progresses, this class may change shape but for current purposes I think itis now ok.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

kutschke commented:
Hi Sophie,

  This is from memory so some details may be wrong but the big picture is right.

1) Your reco module and any data products it makes should have no MC awareness.  The data products should be in RecoDataProducts.

2) In most cases comparison of reco and MC quantities should be done in a separate analyzer module that uses information found in
     data products.

3) It's acceptable for your reco code to have a verbose option that puts extra data products in the event solely for the purpose of 
     comparing to MC.

4) There is another option that you can use if your reco module has transient information that's just too complicated to put into the event.
     art has a concept called a tool, which is a plugin that you can invoke from within a module.  It's OK to write a tool that is called
     from a reco module and which has MC awareness.  For data production will won't ship the tool so the code used on real data
     stays MC free.  Pasha is a local expert on tools.

5) To match a reco track to MC information, navigate back to the digis.  There is a data product of type: StrawDigiMCCollection
    It has the same number of entries as StrawDigiCollection.  To find the MC truth for digi N, look at the Nth entry of StrawDigiMCCollection

6) You can navigate from the StrawDigiMC to the contributing SimParticles.  Remember that many SimParticles may contribute to one digi.
    Plan for this in your design.

7) You can design a summary class that maps reco tracks to SimParticles along with quality of match information.  There are several
    ways to do this. In any case that class should live in MCDataProducts.  It's OK for a class in MCDataProducts to include a header
    from RecoDataProducts.

Having said all of this, when your code is in early to mid development it's OK if your reco algorithms reference MC information.  The
description above is the target as your code approaches maturity.  Keep it in mind so that it's not too hard to make your code
compliant when the time comes.

The story I gave above is for the tracker - we want all subsystems to follow this model.

  Rob



> On Oct 29, 2019, at 11:44 AM, Sophie Middleton <notifications@github.com> wrote:
> 
> Hi,
> 
> Well I guess guidance is best now. I still have the truth info. I have created a class "CosmicTrackTruthInfo" its essentially a struct with just a few members, it just stores: true parameters, true direction, true position vectors. This is only used in the CosmicAnalyzer analysis and is not called or implemented in any fitting routine. The CosmicAnalyzer at the moment is just a diagnostics framework. The CosmicTrack.hh is not totally separte from any truth information. Eventually the CosmicTrackTruthInfo will not be needed. It probably doesnt even need to be a separate file as its just this one struct. I could just place it inside the CosmicAnalyzer file but for now it is its own entity as is place inside my Cosmic package, not in recodataproducts.
> 
> I hope this is ok for now. This whole framework is evolving, I think as I move on to alignment I will make more and more changes but at the moment I just wanted the current "working" imple,ntation to be pulled as requested.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

sophiemiddleton commented:
Hi,

I think my code now meets these requirements. 
1)  Only the analyzer now sees the MC, the truth info is only for validation currently. The Reco code does nothing with the MC.

2) See above, the only difference is that this CosmicTrackInfoMC is not a dataproduct, I mean its just a list of diagnostics so I dont think it needs to be anything more complicated than it is.

3) The reco code doesnt touch MC anymore

4) Its not complicated I think this implementation is not needed .....yet
     
5) This is done already I think. It is done inside the CosmicTrackFinder.....this has an option to store this info or not if the mcdiag is set to "on"

6) Yes I use simparticles - this is all done in the analyzer, hopefully thats ok for now.

7) This may be needed ....prehaps once this reaches maturity.

I will push the code in a minute. All that now needs fixing is to turn the fcl to validation type....I aim to do this this evening. I have to work on calorimeter calibration stuff this afternoon.

sophiemiddleton commented on CosmicReco/src/CosmicAnalyzer_module.cc:
```diff
@@ -0,0 +1,1028 @@
+
+    CosmicAnalyzer::CosmicAnalyzer(fhicl::ParameterSet const& pset) :
+	art::EDAnalyzer(pset),
+	_diag		(pset.get<int>("diagLevel",1)),
```
I have some issues here with defining the SimPArticleTimeOffset. I have it in the Config:

struct Config{
	      using Name=fhicl::Name;
	      using Comment=fhicl::Comment;
	      fhicl::Atom<int> _diag{Name("diagLevel"), Comment("set to 1 for info"),1};
	      fhicl::Atom<bool> _mcdiag{Name("mcdiag"), Comment("set on for MC info"),true};
	      fhicl::Atom<art::InputTag> _chtag{Name("ComboHitCollection"),Comment("tag for combo hit collection")};
	      fhicl::Atom<art::InputTag> _tctag{Name("TimeClusterCollection"),Comment("tag for time cluster collection")};
	      fhicl::Atom<art::InputTag> _costag{Name("CosmicTrackSeedCollection"),Comment("tag for cosmci track seed collection")};
	      fhicl::Atom<art::InputTag> _mcdigistag{Name("StrawDigiMCCollection"),Comment("StrawDigi collection tag"),"makeSD"};
	      fhicl::Atom<fhicl::ParameterSet> _toff{Name("toff"), Comment("Sim particle time offset ")};
	    };

In the initializer I set:

SimParticleTimeOffset _toff;

and then in my findData:

_toff.updatemaps(_evt) 

this causes a massive list of errors....I guess its something I've miss defined but cant see it....any ideas?

The starting error which seems to initialize a cascade of subsequent errors : "error: no matching function for call to 'mu2e::SimParticleTimeOffset::SimParticleTimeOffset()'
         _conf(conf()){}"

gaponenko commented:
Hi Sophie,

The SimParticleTimeOffset class supports validated fhicl, the ParameterSet 
interface is only for compatibility with old code that has not migrated 
yet.   Instead of Atom<ParameterSet> please use  
Atom<SimParticleTimeOffset::Config>.   There is an example in
CaloMC/src/CaloShowerStepROFromShowerStep_module.cc

Andrei


On Tue, 29 Oct 2019, Sophie Middleton wrote:

> sophiemiddleton commented on this pull request.
> 
> 
> 
> > +	
> +      //Flags:
> +	Bool_t _StraightTrackInit, _StraightTrackConverged, _StraightTrackOK, _hitsOK;
> +      //offsets for MC
> +      SimParticleTimeOffset _toff;
> +
> +      Int_t _strawid; // strawid info
> +
> +      vector<ComboHitInfoMC> _chinfomc;
> +
> +      bool findData(const art::Event& evt);
> +    };
> +
> +    CosmicAnalyzer::CosmicAnalyzer(fhicl::ParameterSet const& pset) :
> +	art::EDAnalyzer(pset),
> +	_diag		(pset.get<int>("diagLevel",1)),
> 
> I have some issues here with defining the SimPArticleTimeOffset. I have it in the Config:
> 
> struct Config{
> 	      using Name=fhicl::Name;
> 	      using Comment=fhicl::Comment;
> 	      fhicl::Atom<int> _diag{Name("diagLevel"), Comment("set to 1 for info"),1};
> 	      fhicl::Atom<bool> _mcdiag{Name("mcdiag"), Comment("set on for MC info"),true};
> 	      fhicl::Atom<art::InputTag> _chtag{Name("ComboHitCollection"),Comment("tag for combo hit collection")};
> 	      fhicl::Atom<art::InputTag> _tctag{Name("TimeClusterCollection"),Comment("tag for time cluster collection")};
> 	      fhicl::Atom<art::InputTag> _costag{Name("CosmicTrackSeedCollection"),Comment("tag for cosmci track seed collection")};
> 	      fhicl::Atom<art::InputTag> _mcdigistag{Name("StrawDigiMCCollection"),Comment("StrawDigi collection tag"),"makeSD"};
> 	      fhicl::Atom<fhicl::ParameterSet> _toff{Name("toff"), Comment("Sim particle time offset ")};
> 	    };
> 
> In the initializer I set:
> 
> SimParticleTimeOffset _toff;
> 
> and then in my findData:
> 
> _toff.updatemaps(_evt) 
> 
> this causes a massive list of errors....I guess its something I've miss defined but cant see it....any ideas?
> 
> The starting error which seems to initialize a cascade of subsequent errors : "error: no matching function for call to 'mu2e::SimParticleTimeOffset::SimParticleTimeOffset()'
>          _conf(conf()){}"
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_40-23discussion-5Fr340285503&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=CmqI-OYQ0fNpDs6atAn9398205irlyIWy2V6JO9n-HI&s=IAhx8OK-fhYPAU_2stXRCv8o0devFjfbvcn3mej6nDM&e=

sophiemiddleton commented:
Hi, thanks, it now builds but when I try to run it I get this error:

---- Configuration BEGIN
  
  ====================================================================================================
  
  !! The following modules have been misconfigured: !!
  
  ----------------------------------------------------------------------------------------------------
  
  
  Module label: CosmicAnalysis
  module_type : CosmicAnalyzer
  
  Any parameters prefaced with '#' are optional.
  Unsupported parameters:
  
   + TimeOffsets                    [ /mu2e/app/users/sophie/Offline_October/Offline/CosmicReco/fcl/prolog.fcl:66 ]
   + TimeOffsets.inputs             [  ]
   + TimeOffsets.inputs[0]          [  ]
   + TimeOffsets.inputs[1]          [  ]
  
  
  ====================================================================================================
  
---- Configuration END

In my prolog I have set: TimeOffsets : {inputs : [ @sequence::DigiCompression.NoCosmicTimeMaps ]}

as I am using the MDC cosmics.

sophiemiddleton commented:
hmmm...actually I think this is a result of the validation working, perhaps a typo....I will investigate

gaponenko commented:
Try 

   mu2e --print-descriptint YourModuleName

to see the FCL syntax that it expects.

Andrei

On Tue, 29 Oct 2019, Sophie Middleton wrote:

> Hi, thanks, it now builds but when I try to run it I get this error:
> 
> ---- Configuration BEGIN
>   
>   ====================================================================================================
>   
>   !! The following modules have been misconfigured: !!
>   
>   ----------------------------------------------------------------------------------------------------
>   
>   
>   Module label: CosmicAnalysis
>   module_type : CosmicAnalyzer
>   
>   Any parameters prefaced with '#' are optional.
>   Unsupported parameters:
>   
>    + TimeOffsets                    [ /mu2e/app/users/sophie/Offline_October/Offline/CosmicReco/fcl/prolog.fcl:66 ]
>    + TimeOffsets.inputs             [  ]
>    + TimeOffsets.inputs[0]          [  ]
>    + TimeOffsets.inputs[1]          [  ]
>   
>   
>   ====================================================================================================
>   
> ---- Configuration END
> 
> In my prolog I have set: TimeOffsets : {inputs : [ @sequence::DigiCompression.NoCosmicTimeMaps ]}
> 
> as I am using the MDC cosmics.
> 
> 
> -- 
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub:
> https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Mu2e_Offline_pull_40-23issuecomment-2D547634184&d=DwICaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=O47fc5vzDTR2V_gla4Ub0Q&m=-lIxGeuYjH40SokPwNAJeCY4xY2WLnuKk7a4GYYOjno&s=q2vngtdYHVY2wuqLXFCpAKzqu0MAOTGXWBA25-8DMTY&e=

sophiemiddleton commented:
yes, I fixed that, just a capitol letter missing. I now have a different error : MSG-s ArtException:  PostEndJob 29-Oct-2019 16:38:35 CDT ModuleEndJob
cet::exception caught in art
---- OtherArt BEGIN
  ---- EventProcessorFailure BEGIN
    EventProcessor: an exception occurred during current event processing
    ---- EventProcessorFailure BEGIN
      EndPathExecutor: an exception occurred during current event processing
      ---- ScheduleExecutionFailure BEGIN
        Path: ProcessingStopped.
        ---- ProductNotFound BEGIN
          getByLabel: Found zero products matching all criteria
          Looking for type: mu2e::ComboHitCollection
          Looking for module label: 
          Looking for productInstanceName: 
          cet::exception going through module CosmicAnalyzer/CosmicAnalysis run: 1002 subRun: 358 event: 22
        ---- ProductNotFound END
        Exception going through path end_path
      ---- ScheduleExecutionFailure END
    ---- EventProcessorFailure END
  ---- EventProcessorFailure END
---- OtherArt END
%MSG

again will investigate....

kutschke commented:
Hi Sophie,

  See below

> On Oct 29, 2019, at 4:42 PM, Sophie Middleton <notifications@github.com> wrote:
> 
> yes, I fixed that, just a capitol letter missing. I now have a different error : MSG-s ArtException: PostEndJob 29-Oct-2019 16:38:35 CDT ModuleEndJob
> cet::exception caught in art
> ---- OtherArt BEGIN
> ---- EventProcessorFailure BEGIN
> EventProcessor: an exception occurred during current event processing
> ---- EventProcessorFailure BEGIN
> EndPathExecutor: an exception occurred during current event processing
> ---- ScheduleExecutionFailure BEGIN
> Path: ProcessingStopped.
> ---- ProductNotFound BEGIN
> getByLabel: Found zero products matching all criteria
> Looking for type: mu2e::ComboHitCollection
> Looking for module label:
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

This is the key part.   You asked art to find a data product whose module label
is the empty string.  So that part of your fcl is mixed up.

> Looking for productInstanceName:

Depending on what you are looking for, it may be OK that the instance name is the empty string.
We have many data products with that.


> cet::exception going through module CosmicAnalyzer/CosmicAnalysis run: 1002 subRun: 358 event: 22

This tells you what module and module label threw the exception.

> ---- ProductNotFound END
> Exception going through path end_path
> ---- ScheduleExecutionFailure END
> ---- EventProcessorFailure END
> ---- EventProcessorFailure END
> ---- OtherArt END
> %MSG
> 
> again will investigate....
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

sophiemiddleton commented:
Yeah, I think perhaps its tracks that failed a cut, this is the firstevent and it fails on this so I cant say for sure that it can pass any event. It complains about Combohitcollection but that collection is filled early, perhaps its the new flag system failing and being turned off so something is not filled. I'm not sure yet but hopefully will figure it shortly.

sophiemiddleton commented:
yeah I think this could be an error in a preceding module i.e the producer. I have only transformed the one file into validate fcl currently, this analyzer relies on producers which still use old versions of fcl i.e no validation...some of those are mine i.e CosmicTrackFinder but others are not mine. Perhaps switching over my producer will catch this error. I wanted to debug module at a time to avoid massive incomprehensible errors but maybe thats my issue?

sophiemiddleton commented:
Is there perhaps some way of event dumping here? I have tried command line calls for various things but I can't seem to figure this out. It seems to fail at the end of the first event with just this error message. I've not changed anything in the producer and I now believe this is not a result of the producer but actually inside this new validated analyzer. The analyzer never even makes it into the constructor call. Perhaps the collections are not passed to this point correctly. I've pushed the latest version of the analyzer (above this comment). I'm going to re-read through the wiki pages again and check I've done everything correctly in there. I can't see the difference between my way of constructing the analyzer and the other examples but I guess I use different dataproducts. I will look into solving this after re-reading.

sophiemiddleton commented:
ah, ok, fixed it, I was using the wrong name in my findData()....I can run the fcls and plot outputs.

sophiemiddleton commented:
For my CosmicTrackFit, this was initialized in the old version of CosmicTrackFinder as:
 _tfit        (pset.get<fhicl::ParameterSet>("CosmicTrackFit",fhicl::ParameterSet())). My helper fit function then takes in ParameterSet as an argument and does all the fitting (like HelixSeedFit for example).
 
  I'm not sure how to upgrade this now that ParameterSet is not supported by validated fcl? I dont see that any of the other TrkPatRec code has been transferred over, I also cant find any information on how ParameterSet is now used in the new version of fhichl., I think it is no longer supported.

I also am having some issues as I use art:producttoken and I'm not sure how to translate that over too, again I cant see any documentation on this. For example: I used to have this in my constructor _chToken{consumes<ComboHitCollection>(pset.get<art::InputTag>("ComboHitCollection"))},

I tried a few versions but I keep getting errors thrown.

 I read through kyle's presentation on the changes but didnt see much on either issue. I think this is almost there once I fix these two problems. I just have to translate the helper class and the filter but they should be straight forward as all members are standard c++, I think.

kutschke commented:
Hi Sophie,

  I am reading between the lines a little here so maybe my comment is not on point ...

  In almost all cases we request that if a producer module cannot make any output objects
(ComboHits, Tracks, Clusters .... ), it should put an empty data product into the event
(presuming that the data product is of collection-type).  This simplifies downstream code.
There are rare exceptions in which we will approve of a module not making the data product.

   If you are dealing with one of the exception cases for which it is OK for a data product
to be absent, get products from the event using getByLabel rather than getValidHandle.
Has anyone told you the difference between the two?

Rob
 

  

> On Oct 29, 2019, at 5:45 PM, Sophie Middleton <notifications@github.com> wrote:
> 
> Yeah, I think perhaps its tracks that failed a cut, this is the firstevent and it fails on this so I cant say for sure that it can pass any event. It complains about Combohitcollection but that collection is filled early, perhaps its the new flag system failing and being turned off so something is not filled. I'm not sure yet but hopefully will figure it shortly.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

kutschke commented:
> On Oct 29, 2019, at 11:47 PM, Sophie Middleton <notifications@github.com> wrote:
> 
> For my CosmicTrackFit, this was initialized in the old version of CosmicTrackFinder as:
> _tfit (pset.getfhicl::ParameterSet("CosmicTrackFit",fhicl::ParameterSet())). My helper fit function then takes in ParameterSet as an argument and does all the fitting (like HelixSeedFit for example).
> 
> I'm not sure how to upgrade this now that ParameterSet is not supported by validated fcl? I dont see that any of the other TrkPatRec code has been transferred over, I also cant find any information on how ParameterSet is now used in the new version of fhichl., I think it is no longer supported.

My advice is to ask Andrei again - he recently upgrade a lot of modules to use validated fcl so his memory will be fresher than mine.

> 
> I also am having some issues as I use art:producttoken and I'm not sure how to translate that over too, again I cant see any documentation on this. For example: I used to have this in my constructor _chToken{consumes(pset.getart::InputTag("ComboHitCollection"))},

Have you read:

https://cdcvs.fnal.gov/redmine/projects/art/wiki/Declaring_products_to_consume

I am pretty sure that,in your example, ComboHitCollection should be a template argument, since it is a type, not a string which
is used to represent an InputTag.

> 
> I tried a few versions but I keep getting errors thrown.
> 
> I read through kyle's presentation on the changes but didnt see much on either issue. I think this is almost there once I fix these two problems. I just have to translate the helper class and the filter but they should be straight forward as all members are standard c++, I think.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or unsubscribe.
>

sophiemiddleton commented:
Hi thanks, so I think the empty data products has been solved, at least in the case I described. My main issue is now trying to translate the ParameterSet argument which is handed to the helper fit function. I also need to fix those product tokens (see previous comment). I have emailed Andrei but will also do some of my own reading to gain a better understanding here. Thanks, Sophie

bonventre commented:
Hi Sophie, just a couple random comments: 
1. Should DriftFitUtils.hh be in the CosmicReco directory now too?
2. I think some of the class naming could be reworked to be more specific, like FullFit and DataFit, and possibly moved/combined. Those two for example I think are only separate in my version because I have separate fits for the prototype data and the offline sims, but you should only need one version of the fit.
3. Similarly does LikelihoodFunctions need to be separate from CosmicTrackFit?
4. Maybe this should just be left alone in case we want the functionality, but in your version of full fit since you are fixing sigma you should be able to use a 2D table for the PDF instead of 3D

sophiemiddleton commented:
Hi, thanks:
1) Dave said to move to TrackerConditons for future use as it will need to use that code
2)
3) I dont know, I guess it kind of sits in its own namepsace, I like that its sperate as this code might be used for other Cosmic fittiing, I guess the Liklihood function is quite tracker specific.....I dont know. I'll think it through

sophiemiddleton commented:
Richie, looking at the code, I will rename DataFit ---->PDFFitter, TimePDF---->GaussianPDF, FullFit at the moment I've call FullDriftFit....I guess additonal/alternative PDFs would then be add ons of PDFFitter ....I think that makes sense....I will spend some time thinking it over....

sophiemiddleton commented:
Ok, I've pushed the code. This latest version has all of Dave's revisions and some changes on Richie's request. Please let me know if further revisions are necessary. This is still early days with this study and this is just the first step in this CosmicReco analysis so it will evolve from this so a few of the things you have mentioned will of course migrate to a more sophisticated framework as we reach maturity here.  Saying that, I think for now it should, hopefully, fulfill requirements. But of course let me know if I need to make further changes.

Thanks,
Sophie

brownd1978 commented:
Hi Sophie,
thanks for taking the time to address the review requests.  I think the
changes are positive, I hope you do to.  I agree progress can be made
incrementally from here via normal development.  I plan to merge your pull
tomorrow morning, so we don't confuse them with the changes merged today.
cheers, Dave

On Wed, Oct 30, 2019 at 8:28 PM Sophie Middleton <notifications@github.com>
wrote:

> Ok, I've pushed the code. This latest version has all of Dave's revisions
> and some changes on Richie's request. Please let me know if further
> revisions are necessary. This is still early days with this study and this
> is just the first step in this CosmicReco analysis so it will evolve from
> this so a few of the things you have mentioned will of course migrate to a
> more sophisticated framework as we reach maturity here. Saying that, I
> think for now it should, hopefully, fulfill requirements. But of course let
> me know if I need to make further changes.
>
> Thanks,
> Sophie
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Mu2e/Offline/pull/40?email_source=notifications&email_token=ABAH57Y2C2MYCW74IJBYMODQRJGFHA5CNFSM4JFGSWLKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECWOLWQ#issuecomment-548201946>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABAH576ZB6BM657RV7I4JGDQRJGFHANCNFSM4JFGSWLA>
> .
>


-- 
David Nathan Brown            Dave_Brown@lbl.gov
Office Phone (510) 486-7261  Fax  495-2957
Lawrence Berkeley National Lab
MS 50R5008 (50-6026C)  Berkeley, CA 94720

sophiemiddleton commented:
excellent, thanks, yes the code is a lot more streamlined now since your review. I think it had a acquired a lot of unnecessary features over the course of its intermittent development and when I had time to read it in its entirety I could see where it had to improve. Cheers, Sophie

sophiemiddleton commented:
Hi, yeah I was using IsValid - perhaps I removed that now? I think I then decided that sometimes it can be valid i.e .converged but still not a good fit so I switched to having some cut on maximum LogL...as I said this part of the code is likely to be changed as I move towards alignment/other uses. I think you are right that likelihood should eventually be part of CosmicTrackFit but as it is still being worked on I will set aside for now. I will also think over that class structure once I work on this again for alignment purposes. Once I have the alignment framework setup the migration of these things will become necessary. 
Thanks,Sophie

sophiemiddleton commented:
Hi, Ok thats far enough.

1) easily done.

These will be more work:

2) Ok, I think for easiest sake a rename for now as perhaps I should call PDFFitter FullDriftFit and then FullDriftFit CombinedFit or something as the FullDriftFit is the endresult and the other function which is now called FullDriftFit is just one version

3) rename again is easier.....perhaps I should call this name space MinuitDriftFit

sophiemiddleton commented:
I have implemented changes 1 and 3 - let me know if ok? Will look at 2 now...

sophiemiddleton commented:
Hi, Ok Richie, I have completed the change for your point 2. I have removed PDFFitter, its fucntion was pointless in this version. I think this is ok now but let me know if further issues.
Thanks,
Sophie

bonventre commented:
I think you forgot to add the renamed likelihoodfunctions file to the git commit!

sophiemiddleton commented:
oops, sorry, added now, cheers, Sophie
